src/webgpu/web_platform/reftests/canvas_size_different_with_back_buffer_size.html.ts

```diff
@@ -0,0 +1,208 @@
+
+import { runRefTest } from './gpu_ref_test.js';
+
+// <canvas> element from html page
+declare const cvs_larger_than_back_buffer: HTMLCanvasElement;
+declare const cvs_same_as_back_buffer: HTMLCanvasElement;
+declare const cvs_smaller_than_back_buffer: HTMLCanvasElement;
+
+export function run() {
+  runRefTest(async t => {
+    const red = new Uint8Array([0x00, 0x00, 0xff, 0xff]);
+    const green = new Uint8Array([0x00, 0xff, 0x00, 0xff]);
+    const blue = new Uint8Array([0xff, 0x00, 0x00, 0xff]);
+    const yellow = new Uint8Array([0x00, 0xff, 0xff, 0xff]);
+
+    {
+      const ctx = cvs_larger_than_back_buffer.getContext('gpupresent');
+      assert(ctx !== null);
+      const width = cvs_larger_than_back_buffer.width / 2;
+      const height = cvs_larger_than_back_buffer.height / 2;
+      ctx.configure({
+        device: t.device,
+        format: 'bgra8unorm',
+        usage: GPUTextureUsage.COPY_DST,
+        size: [width, height, 1], // half of the canvas width and height.
+      });
+
+      const rows = height;
+      const bytesPerRow = 256;
+      const buffer = t.device.createBuffer({
+        mappedAtCreation: true,
+        size: rows * bytesPerRow,
+        usage: GPUBufferUsage.COPY_SRC,
+      });
+      const mapping = buffer.getMappedRange();
+      const data = new Uint8Array(mapping);
+      // 1st row
+      data.set(red, 0);
+      data.set(red, 4);
+      data.set(green, 8);
+
+      // 2nd row
+      data.set(red, 256 + 0);
+      data.set(red, 256 + 4);
+      data.set(green, 256 + 8);
+
+      // 3rd row
+      data.set(blue, 256 * 2 + 0);
+      data.set(blue, 256 * 2 + 4);
+      data.set(yellow, 256 * 2 + 8);
+
+      // 4th row
+      data.set(blue, 256 * 3 + 0);
+      data.set(blue, 256 * 3 + 4);
+      data.set(yellow, 256 * 3 + 8);
+      buffer.unmap();
+
+      const texture = ctx.getCurrentTexture();
+
+      const encoder = t.device.createCommandEncoder();
+      encoder.copyBufferToTexture({ buffer, bytesPerRow }, { texture }, [width, height, 1]);
+      t.device.queue.submit([encoder.finish()]);
+      await t.device.queue.onSubmittedWorkDone();
+    }
+
+    {
+      const ctx = cvs_same_as_back_buffer.getContext('gpupresent');
+      assert(ctx !== null);
+      ctx.configure({
+        device: t.device,
+        format: 'bgra8unorm',
+        usage: GPUTextureUsage.COPY_DST,
+      });
+
+      const rows = cvs_same_as_back_buffer.height;
+      const bytesPerRow = 256;
+      const buffer = t.device.createBuffer({
+        mappedAtCreation: true,
+        size: rows * bytesPerRow,
+        usage: GPUBufferUsage.COPY_SRC,
+      });
+      const mapping = buffer.getMappedRange();
+      const data = new Uint8Array(mapping);
+      // 1st row
+      data.set(red, 0); // red
+      data.set(red, 4); // red
+      data.set(green, 8); // green
+
+      // 2nd row
+      data.set(red, 256 + 0); // red
+      data.set(red, 256 + 4); // red
+      data.set(green, 256 + 8); // blue
```

Comment:
I think you can factor out a small helper so each canvas can have something as simple as:
```ts
setRowColors(data, 256 * 0, [red, red, green]);
setRowColors(data, 256 * 1, [red, red, green]);
setRowColors(data, 256 * 2, [blue, blue, yellow]);
setRowColors(data, 256 * 3, [blue, blue, yellow]);
```

That way these three cases will be way less repetitive.

Reply:
Thanks for the suggestion.